prometheus/converters.go

```diff
@@ -0,0 +1,175 @@
+package prometheus
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/influxdata/influxdb/models"
+	"github.com/prometheus/prometheus/storage/remote"
+)
+
+const (
+	// measurementName is where all prometheus time series go to
+	measurementName = "_"
+
+	// fieldName is the field all prometheus values get written to
+	fieldName = "f64"
+)
+
+var ErrNaNDropped = errors.New("dropped NaN from Prometheus since they are not supported")
+
+// WriteRequestToPoints converts a Prometheus remote write request of time series and their
+// samples into Points that can be written into Influx
+func WriteRequestToPoints(req *remote.WriteRequest) ([]models.Point, error) {
+	var maxPoints int
+	for _, ts := range req.Timeseries {
+		maxPoints += len(ts.Samples)
+	}
+	points := make([]models.Point, 0, maxPoints)
+
+	var droppedNaN error
+
+	for _, ts := range req.Timeseries {
+		tags := make(map[string]string, len(ts.Labels))
+		for _, l := range ts.Labels {
+			tags[l.Name] = l.Value
+		}
+
+		for _, s := range ts.Samples {
+			// skip NaN values, which are valid in Prometheus
+			if math.IsNaN(s.Value) {
+				droppedNaN = ErrNaNDropped
+				continue
+			}
+
+			// convert and append
+			t := time.Unix(0, s.TimestampMs*int64(time.Millisecond))
+			fields := map[string]interface{}{fieldName: s.Value}
+			p, err := models.NewPoint(measurementName, models.NewTags(tags), fields, t)
+			if err != nil {
+				return nil, err
+			}
+
+			points = append(points, p)
+		}
+	}
+	return points, droppedNaN
+}
+
+// ReadRequestToInfluxQLQuery converts a Prometheus remote read request to an equivalent InfluxQL
+// query that will return the requested data when executed
+func ReadRequestToInfluxQLQuery(req *remote.ReadRequest, db, rp string) (*influxql.Query, error) {
+	if len(req.Queries) != 1 {
+		return nil, errors.New("Prometheus read endpoint currently only supports one query at a time")
+	}
+	promQuery := req.Queries[0]
+
+	q := &influxql.Query{}
```

Comment:
It would probably be easier to just create this at the bottom when you need it with a literal. The literal will likely allocate less memory to boot since it doesn't use `append`.